framework/src/bcs/MatNeumannBC.C

```diff
@@ -0,0 +1,42 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MatNeumannBC.h"
+
+registerMooseObject("MooseApp", MatNeumannBC);
+
+defineLegacyParams(MatNeumannBC);
+
+InputParameters
+MatNeumannBC::validParams()
+{
+  InputParameters params = IntegratedBC::validParams();
+  params.addParam<Real>("value", 0.0, "The value to be enforced on the boundary.");
+  params.declareControllable("value");
+  params.addClassDescription("Imposes the integrated boundary condition "
+                             "$\\frac{D \\partial u}{\\partial n}=M*h$, "
```

Comment:
I noticed that you added a `D` compared to the description of `NeumannBC`. That's only correct when applying this BC to diffusion type PDEs and I `D` doesn't show up in this kernel at all. I don't like the `NeumannBC` explanation either, but maybe we can find a better wording that will avoid the confusion we had around this BC.

Reply:
Yeah, the reason I did that was that a couple of weeks ago, for the SPS project, the existing `NeumannBC` documentation led me into some confusion. I was wanting to use a `NeumannBC` to enforce the gradient of the variable to have a certain value, rather than trying to enforce a certain value of the flux. So, not having worked with Neumann BC's before, I read the documentation and thought I just needed to specify the value I wanted the gradient to have. But, I had a non-unity diffusion coefficient, so I was not getting the value I expected. It took a bit of time and head-scratching to figure out that it is actually enforcing D*grad u at the boundary and I hadn't accounted for the non-unity D. Admittedly this is largely due to me just not thinking the problem through carefully, but I feel like the existing documentation could lead to similar confusion for others. I don't know what other types of problems this BC is typically applied to, but it seems like diffusion-type problems are going to be among the most common applications, so to me it makes sense to include the `D` in the class description and documentation. But, I'm open to other ideas. We could use a constant other than `D` that is less specific to diffusion problems, for instance.

Reply:
@laagesen Did you account for this by simply dividing your desired gradient by the non-unity `D`? 

My two cents on naming: `D` comes up often in my electromagnetics/electrostatics models as multiple different properties: `inverse_mu`, `sigma`, `epsilon`, etc. etc. So giving a coefficient letter that's not `D` or any of these others is probably best to keep it general. As far as a description, maybe something like:

```
Imposes the integrated boundary condition $\\frac{C \\partial u}{\\partial n}=M*h$, where $h$ is a constant, $M$ is a material property, and $C$ is a coefficient of the Laplacian-style operator ($\\nabla \\cdot (C \\nabla u)$).
```

Going wordier here might not be best, but I think it does leave things clearer than previous (at least to me).

Reply:
@cticenhour thanks for the input. I like C since it seems a little more generic than D, and will revise based on that. And yes, I ended up just dividing by the non-unity `D` to get the gradient I wanted.